### PR TITLE
feat: add `foreach` expression

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to #3950

### CODE REVIEW

File: main/test/flix/Test.Exp.Foreach.flix [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace Test/Exp/Foreach {
+
+    @test
+    def testForeach01(): Int32 = // Dummy program that just checks compiler phases
+        foreach (x <- xs) x + 1
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.Foreach.flix:5]: Now you can write some test cases to experiment with what you think should be allowed :)

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 2]
```diff
@@ -699,6 +699,10 @@ object Weeder {
         case e => WeededAst.Expression.Discard(e, loc)
       }
 
+    case ParsedAst.Expression.ForEach(sp1, pat, exp1, exp2, sp2) =>
+      val loc = mkSL(sp1, sp2)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:703]: For simplicity, you can do an unsafe pattern match on `pat` to only handle the case where it is a variable. E.g. something like:

```
val varSym = pat match {
  case Pattern.Var(...) => ..
  case _ => ???
```

Then you recursively visit exp1 and exp2.

Then you need to construct a call `List.foreach(varSym -> exp1, exp2)` which means you need to construct some applies nodes and a Def node.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 3]
```diff
@@ -704,6 +704,17 @@ object ParsedAst {
...
+    /**
+      * ForEach Expression.
+      *
+      * @param sp1     the position of the first character in the expression.
+      * @param pat     the pattern on the left hand side.
+      * @param genExp  the generator expression.
+      * @param bodyExp the body expression.
+      * @param sp2     the position of the last character in the expression.
+      */
+    case class ForEach(sp1: SourcePosition, pat: ParsedAst.Pattern, genExp: ParsedAst.Expression, bodyExp: ParsedAst.Expression, sp2: SourcePosition) extends ParsedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:716]: By convention we just call these exp1 and exp2

**Comment by magnus-madsen** [comment]
Very cool!

File: main/test/flix/Test.Exp.Foreach.flix [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        let z = ref 0 @ r;
+        let q = ref 1 @ r;
+        let l = (1, 2) :: (3, 4) :: Nil;
+        foreach ((x, y) <- l) {
+            q := deref q + x;
+            z := deref z + x + y
+        };
+        deref z + deref q == 15
+    }
+
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.Foreach.flix:62]: How about some nested foreach? And how about some more tests to ensure the parsing is as expected?

File: main/test/flix/CompilerSuite.scala [hunk 5]
```diff
@@ -5,6 +5,8 @@ import ca.uwaterloo.flix.util.{FlixSuite, Options}
 class CompilerSuite extends FlixSuite(incremental = true) {
   implicit val options: Options = Options.TestWithLibAll
 
-  mkTestDir("main/test/flix/")
+  // REVERT ME
+  //mkTestDir("main/test/flix/")
```
Comment (magnus-madsen) [inline main/test/flix/CompilerSuite.scala:9]: Dummy code

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 6]
```diff
@@ -1622,6 +1628,21 @@ object Weeder {
       }
   }
 
+  private def mkLambdaMatch(sp1: SourcePosition, p: WeededAst.Pattern, e: WeededAst.Expression, sp2: SourcePosition)(implicit flix: Flix) = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:1631]: Add a comment that explains what this does.

**Comment by jaschdoc** [comment]
RFR

**Pull request merged**